sql/src/main/java/org/apache/druid/sql/calcite/rule/DruidJoinRule.java

```diff
@@ -193,9 +193,16 @@ static boolean canHandleCondition(final RexNode condition, final RelDataType lef
     final int numLeftFields = leftRowType.getFieldCount();
 
     for (RexNode subCondition : subConditions) {
-      if (subCondition.isA(SqlKind.LITERAL)) {
-        // Literals are always OK.
-        literalSubConditions.add((RexLiteral) subCondition);
+      if (RexUtil.isLiteral(subCondition, true)) {
+        if (subCondition instanceof RexCall) {
+          // This is CAST(literal) which is always OK.
+          // We know that this is CASE(literal) as it pass the check from RexUtil.isLiteral
+          RexCall call = (RexCall) subCondition;
+          literalSubConditions.add((RexLiteral) call.operands.get(0));
```

Comment:
We have to verify the types of the cast here, because if the underlying literal and the cast output type are different, then skipping the cast might change the meaning of the subcondition. Try looking at the type of the RexCall and comparing it to the type of the underlying RexLiteral. I think some logic like this would work:

1. If the types are the same, unwrap the cast and use the underlying literal.
2. If the types are not the same, return `Optional.empty()` indicating the condition is not supported.

Alternatively, actually apply the cast in step (2) instead of rejecting the condition. I'm not sure how exactly you would do this or whether it is necessary. If it turns out to be necessary, perhaps check out what ReduceExpressionsRule does.

Reply:
Done